## Prompt Engineering Tutorial!
This tutorial demonstrates how changing the prompt affects the output, illustrating the importance and practical impact of prompt engineering using OpenAI’s API (works similarly for other LLMs like Gemini or Cohere)

### Setup and Configuration

In [13]:
!pip install openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 48.6 MB/s eta 0:00:00


In [14]:
from openai import OpenAI

# 👇 Set your API key here directly
client = OpenAI(api_key="Add Your OpenAI API Key")

### Zero-Shot Prompting
Ask a question without giving examples — this tests the model’s built-in knowledge.

In [15]:
response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "user", "content": "Can you recommend some science-fiction books?"}
    ]
)

print(response.choices[0].message.content)

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"


Certainly! Here are some classic and contemporary science fiction books that span various themes and styles:

### Classics:
1. **"Dune" by Frank Herbert** - A sweeping epic set on the desert planet of Arrakis, focusing on politics, religion, and ecology.
2. **"Neuromancer" by William Gibson** - A foundational cyberpunk novel that explores artificial intelligence and virtual reality.
3. **"The Left Hand of Darkness" by Ursula K. Le Guin** - A thought-provoking tale about gender and society on a planet where inhabitants can change their sex.
4. **"Fahrenheit 451" by Ray Bradbury** - A dystopian novel about a future society where books are banned and "firemen" burn them.
5. **"Foundation" by Isaac Asimov** - A groundbreaking series that follows the fall and rise of civilizations through the lens of psychohistory.

### Contemporary:
6. **"The Three-Body Problem" by Liu Cixin** - A hard science fiction novel that explores first contact with an alien civilization and the complexities of phys

### Few-Shot Prompting
Give examples to show the model what pattern to follow.

In [16]:
prompt = """
Convert the following temperatures from Celsius to Fahrenheit:
0°C → 32°F
25°C → 77°F
-10°C → 14°F
100°C →
"""

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "user", "content": prompt}
    ]
)

print(response.choices[0].message.content)

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"


To convert temperatures from Celsius to Fahrenheit, you can use the formula:

\[ F = \frac{9}{5}C + 32 \]

Now, let's apply this formula to convert 100°C to Fahrenheit:

\[ F = \frac{9}{5}(100) + 32 \]
\[ F = 180 + 32 \]
\[ F = 212°F \]

So, 100°C is equal to 212°F.


The model understands the pattern and predicts the missing conversion (212°F).

### Context and Constraints
Control role, tone, and structure of the output.

In [18]:
response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "system", "content": "You are an expert Python developer."},
        {"role": "user", "content": "Write a clear Python function to calculate factorial, including short inline comments."}
    ]
)

print(response.choices[0].message.content)

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"


Certainly! Below is a Python function that calculates the factorial of a non-negative integer using a simple iterative approach. I've included inline comments for clarity.

```python
def factorial(n):
    """
    Calculate the factorial of a non-negative integer n.

    :param n: Non-negative integer
    :return: Factorial of n
    """
    # Check if the input is a negative integer
    if n < 0:
        raise ValueError("Input must be a non-negative integer.")
    
    # Initialize the result to 1 (0! = 1)
    result = 1

    # Iterate from 1 to n to compute the factorial
    for i in range(1, n + 1):
        result *= i  # Multiply the current result by i

    return result  # Return the final computed factorial
```

### Example Usage:
```python
print(factorial(5))  # Output: 120
print(factorial(0))  # Output: 1
```

In this function:
- We first check if the input `n` is a non-negative integer and raise a `ValueError` if it's not.
- We initialize a variable `result` to 1, since the fa

### Chain-of-Thought Prompting 
Encourage the model to reason step-by-step before giving an answer.
(You’re not exposing the reasoning; you’re prompting it to think.)

In [19]:
prompt = """
A student scored 72, 85, and 90 in three subjects.
What is the average score?
Let's think step-by-step before answering.
"""

response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[{"role": "user", "content": prompt}]
)
print(response.choices[0].message.content)

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"


To find the average score of the student in the three subjects, we need to follow these steps:

1. **Add the scores together**: We need to sum the scores of the three subjects.
   \[
   72 + 85 + 90
   \]

   Performing the addition:
   \[
   72 + 85 = 157
   \]
   \[
   157 + 90 = 247
   \]

   So, the total score is \( 247 \).

2. **Count the number of subjects**: In this case, the student has scores in 3 subjects.

3. **Calculate the average**: The average score is obtained by dividing the total score by the number of subjects.
   \[
   \text{Average} = \frac{\text{Total Score}}{\text{Number of Subjects}} = \frac{247}{3}
   \]

   Now, performing the division:
   \[
   247 \div 3 = 82.33 \, (\text{approximately})
   \]

Thus, the average score of the student is approximately **82.33**.


### Prompt Chaining
Use the output of one prompt as the input to another.
This simulates multi-step reasoning or a mini-workflow.

In [20]:
# Step 1: Generate a short article
article_response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[{"role": "user", "content": "Write a short paragraph on the importance of renewable energy."}]
)
article = article_response.choices[0].message.content
print("Generated Article:\n", article)

# Step 2: Summarize the generated article
summary_response = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "user", "content": f"Summarize this paragraph in one sentence:\n{article}"}
    ]
)
print("\nSummary:\n", summary_response.choices[0].message.content)

INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"


Generated Article:
 Renewable energy is crucial for a sustainable future as it offers a viable alternative to fossil fuels, significantly reducing greenhouse gas emissions and mitigating climate change. By harnessing natural resources such as sunlight, wind, and water, renewable energy sources promote energy independence and security, creating resilient economies. Furthermore, the transition to renewable energy supports job creation in emerging sectors, stimulates technological innovation, and enhances public health by reducing air pollution. Embracing renewable energy not only addresses environmental challenges but also fosters a sustainable and equitable society for future generations.


INFO:httpx:HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"



Summary:
 Renewable energy is essential for a sustainable future as it reduces greenhouse gas emissions, enhances energy independence, creates jobs, fosters technological innovation, and improves public health, while addressing environmental challenges and promoting equity.


Prompt Chaining mimics how agentic workflows and RAG systems operate —
passing intermediate outputs through reasoning or transformation steps.